Add support for hot reload of modified code

yoDon opened the pull request:
A slight modification of the hello-world sample that supports hot reloading of modified code while debugging in VSCode (added as a new sample in this PR).

eanders-MS commented:
Hi yoDon. Thanks for this. Is there a reason you duplicated hello-world rather than modifying the existing sample? If there is no downside to this change, I think it would be worth adding it to all the existing samples.

yoDon commented:
Hi eanders, 

The changes I made introduce one slight change in the debugging workflow, which is you need to npm run debug and then attach the debugger, rather than just directly launching the debugger. From my perspective hot-reload is worth incurring that slight change for but I didn't want to presume others would feel the same way.

I messed around a bit with launch.json trying to get the debugger to reconnect properly in the direct launching approach but there seems to be some bit of magic/inconsistency in either VSCode or nodemon that's making that tricky, hence the attaching approach.

I'm happy to rework the PR as a change to the existing samples, that should be a pretty quick change. If I end up reworking it and you end up wanting to stick with the original separate sample approach that's fine too, no worries.

yoDon commented:
Quick update - while looking at adding hot reload to the other samples I caught an additional dependency in the process I need to handle – in the current PR you need to do an npm run build before doing the npm run debug to avoid a race condition between the Typescript compiler finishing first or the Node launcher finishing first. I’ll have a patch for that in an updated PR shortly.

yoDon commented:
@eanders-MS - This commit integrates the code changes into the existing code samples instead of adding a new code sample. I also fixed a small bug in the original commit having to do with the interplay of npm run build and npm run debug, extended the samples to support both async and regular `started()` methods, and added README details to each of the samples.

eanders-MS commented on samples/hello-world/package.json:
```diff
@@ -23,21 +23,25 @@
+    "start-watch": "nodemon --nolazy --inspect .",
     "start": "node .",
-    "debug": "node --nolazy --inspect-brk=9229 ."
+    "debug": "npm run build && concurrently \"npm run *-watch\""
```
In keeping with my comment to keep the launch tasks for the time being, I'd like to keep the "debug" script as-is, and add a new one called "debug-watch" that runs the watch debugger. I think it would be good if this script launched nodemon with the --inspect-brk parameter, avoiding the need for the delay start code in server.ts (see my comment in server.ts about that). But I'm not too familiar with nodemon .. I could be wrong about how it works!

yoDon replied:
I started with exactly what you proposed, and then realized it was causing an infinite loop due to the wildcard in the concurrently run *-watch line. That's totally fixable of course (we could explicitly name the two steps to avoid the wildcard) but it made me think that although it's a new idea to the seasoned old MRE hands, hot-reload debugging seems like it would be the recommended/default preference for 99% of users going forward (the SDK team itself or similar might be the exception). If that's the case, then the debug script should I think use the nodemon path because that's going to be the natural and expected default by both users and toolchains (but that's 100% personal opinion here and obviously I'm 100% fine with any names for these commands)

yoDon commented on samples/hello-world/src/app.ts:
```diff
@@ -31,6 +31,35 @@ export default class HelloWorld {
 	 * Once the context is "started", initialize the app.
 	 */
 	private started() {
+		// Check whether code is running in a debuggable watched filesystem
```
I'm not sure I understand the question here. 

If you mean do we need to do a setTimeout() when debugging, I believe the answer is yes. If we want to be able to set breakpoints in the started() function during hot reload,  we need to avoid the race condition where the start() function runs before the debugger has time to attach. The only way I see to do that is by inserting an artificial delay in the started() function to make sure the debugger wins the race (it may be possible to use some other npm script magic to force a similar delay without changing the code here, but in my experience that stuff is a lot more black-magicy and a lot more opaque to most developers than a simple setTimeout() call so I'd still vote for doing it here).

If you mean did I check whether the debug-detection logic works with --inspect-brk as opposed to --inspect, yes it should work with either (I debated adding an explicit check for "inspect-brk" in addition to checking for "inspect", but since the check is using includes() rather than === eventually some clever person or lint improvement will note that the second check is redundant and drive a PR to remove it).

yoDon replied:
Ah the joys of a race condition :) On my machine (Surfacebook 3 i7), `tsc` and `nodemon` play nice without the timeout but the debugger fails to re-attach on hot reload fast enough to catch a breakpoint in `started()` without the timeout. My thinking was there was enough other reload latency happening that a 1 sec delay (tuneable by the individual developer) wouldn't be a workflow killer and that it allowed support for folks who would be developing on slower machines (or lower memory or whatever secret factor determines the winner of the race condition).

eanders-MS replied:
Did you test this with --inspect-brk passed to nodemon? This *should* made node wait for the debugger to reattach before continuing with app startup.

yoDon replied:
Yes, --inspect-brk behaves as you described. The problem for me is that the "expected" behavior of that flag is not what I personally expected when doing an npm run debug type command. I tend to always kick off the debug variant (which in my world is usually the same as the start variant, since I'm usually in a world of either start or build not start or debug). Bringing that set of npm script expectations to the table led me to think something was broken in my environment when I did an npm run debug-watch and nothing happened. I had to do a fair amount of digging and guessing to realize that I needed to connect the debugger to get it to work (I don't think that's a behavior I've encountered in my dev history, even though it's clearly baked into the node flags so someone must find it expected).

From my perspective that "crap, this thing I just downloaded is not working, I don't know why, I'll give up on this tech because it's probably broken" moment would be an experience I'd optimize the new-user environment to avoid, even if it meant a few extra lines of boilerplate code that might be hard for a novice to understand (but that they could in practice simply leave untouched and be fine). 

I'll follow your lead here, but I'm definitely of the opinion that --inspect-brk is great for the launch config but not so great for the attach variant.

eanders-MS replied:
OK, I've come around to your position on --inspect-brk, but possibly for a different reason. For me, it is distracting that the debugger hits the automatic startup breakpoint upon every hot-reload. I have a commit ready that I will push momentarily. It moves the startup delay to server.ts. Curious how this sits with you.

yoDon replied:
Moving the startup delay into server.ts is a great idea. Total win, correct place for it.

yoDon commented on .vscode/launch.json:
```diff
@@ -6,67 +6,39 @@
   "configurations": [
     {
       "type": "node",
-      "request": "launch",
```
I've updated the PR to include both launch and attach for now. 

That said: 

With hot-reloading, you come in in the morning, `npm run debug` once, attach the debugger once, and you're good for the rest of the day, including surviving through restarts of your AltspaceVR viewer app. That's quite different from the current situation where everyone is used to having to restart their node MRK every time they touch the code. I think keeping hot-reload out of the samples is a huge mistake since it "just works" and anyone who truly needs the old attach behavior is probably an advanced dev doing some kind of specialized stuff for which they are willing to forego hot-reload and for which they can do their own old-school non-nodemon, non-tsc -w, no not-reload config while most everyone else is happier with the hot reloading workflow.

Diving deeper into the details:

Attaching seems to be 100% stable while Launching is fragile and appears to lock some port in a manner that appears to work but starts to fail silently until the machine reboots. When using Launch, the code under test loads properly in AltspaceVR, and even reloads properly after then first edit-and-save cycle, but any future edits fail to load. I've poked around at this a fair amount and can't tell if it's a bug in the VSCode debugger, something in Node or npm or nodemon, or something in the code of the MRE SDK. What I can see is that, when using attach, after the first code change, the code under test reloads correctly in the AltspaceVR viewer but the VSCode debugger console reports `Error: EPIPE: Broken Pipe, write` on server.ts line 13, after which everything appears to be working correctly but AltspaceVR ceases to reload changes until the PC is restarted (closing/relaunching VSCode and AltspaceVR is not enough).

![ErrorMessage](https://user-images.githubusercontent.com/1760705/77784909-9506d380-7018-11ea-810b-a2170242cfd7.JPG)

I've reactivated the old Launch configs so you can test more easily on your end. My recommendation is we keep this commit for testing but merge the attach version into master to benefit new devs while waiting for a launch fix.

yoDon replied:
Also, in the interest of reducing the surface area to explore if anyone does want to dig into this, there are a couple config options in use that are available to launch that are not available to debug. I've temporarily commented them out to highlight them as potentially relevant at some point but apparently not related to the broken pipe blocker.

yoDon replied:
One small note - In the new debug-watch command I'm using --inspect rather than --inspect-brk. I tried using `--inspect-brk`, but the problem is it auto-inserts a breakpoint at the start of execution even if no debugger is attached. In the debug-watch + attach scenario, I found this confusing and I burned a bunch of time trying to debug what was wrong with my code before realizing doh! the code's probably stopped at the auto-breakpoint and I need to attach the debugger in order to be able to hit continue. My view is others are even more likely to think something is wrong (I've learned way more about --inspect-brk in the past couple days than I ever knew before) so I'm advocating we have debug-watch do `--inspect` rather than `--inspect-brk`.

eanders-MS replied:
That seems reasonable (I'm fine with either one). If someone needs to debug app startup, they can change it locally.

yoDon commented on samples/hello-world/README.md:
```diff
@@ -1,15 +1,29 @@
 
-* From inside VSCode: `Shift+Ctrl+B`
-* From command line: `npm run build`
+* Open one of your worlds in AltspaceVR
```
I'm good with any approach you prefer. I found that as someone new to AltspaceVR encountering the MRE SDK for the first time, there was quite a bit that the samples assumed I knew that wasn't obvious. By way of example, I burned quite a bit of time in google and still couldn't find anything that explained where to find the World Editor, or that more specifically explained you needed to enable Early Access to get the World Editor button. I'm good with any edits you want to make to the README but I'll respectfully advocate as someone brand new to the SDK that a detailed walkthrough is actually more helpful than harmful.

yoDon replied:
Doc formatting is a sufficiently personal preference thing that I'm going to recommend I let you take nudge one of the README's into shape first. I'm happy to do the work to bring the others in line to match your sample if you'd like, but the rate of convergence on what you want is likely to be too inefficient if I try to guess at it. Is that too much "passing of the buck" for you? If so I'll go ahead and try to take a first crack at it and you can critique from there.

eanders-MS replied:
Sure, I can do that no problem. Do you mind if I push directly to your branch?

yoDon replied:
Sure, pushing to my branch sounds like an easy way to contain the changes. I'll keep my eye out for it.

eanders-MS replied:
Oh right this is in a fork. If you like, you can give me permissions: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork. Alternatively, I can fork it myself and PR back. Either way, I won't be able to get to this until later in the day/eve today.

eanders-MS replied:
Ok, my changes are in for Hello World. I'd like to know if they land well for you, as a new user who experienced some pain points getting started.

yoDon replied:
Looks really good, and looks like all but the last 4 bullets are sample-independent so easy to reuse. My only quibble is the auto-save highly recommended bit. That quibble is partly personal (with hot reload I often purposely wait and use save or save all like a build command so my app stays usable until I have a complete change ready) and partly generic (when I write tutorials I try to either keep personal opinions out of it or else be highly opinionated and explain the motivation behind the opinions - here, as written, my view is a novice dev wouldn't be able to assess the context behind your recommendation). That's totally a personal stylistic thing though, and you're the best keeper of the vision you want to work the sample docs towards so I'm good with either.

eanders-MS replied:
Your quibble with auto save vs. hot reload is completely warranted. I hadn't thought of it from that angle, but it makes perfect sense. I also try to stick to a non-opinionated (neutral but approachable) voice. The inline recommendation goes against that as well. Thanks for that catch. Will remove.

yoDon commented on samples/hello-world/README.md:
```diff
@@ -1,15 +1,29 @@
-* From inside VSCode: `F5`
-* From command line: `npm start`
+* `npm run debug`
+* Switch to the VSCode debugger panel using the icon at the far left of VSCode that looks like a little bug over a Play button
```
I'm good with any approach you prefer. I found that as someone new to AltspaceVR encountering the MRE SDK for the first time, there was quite a bit that the samples assumed I knew that wasn't obvious. By way of example, I burned quite a bit of time in google and still couldn't find anything that explained where to find the World Editor, or that more specifically explained you needed to enable Early Access to get the World Editor button. As someone brand new to the SDK that a detailed walkthrough is actually more helpful than harmful.

yoDon commented:
(notes on most recent commit are threaded into the conversations above)

yoDon commented on .vscode/launch.json:
```diff
@@ -4,58 +4,100 @@
       "type": "node",
       "request": "launch",
-      "name": "Launch tic-tac-toe project",
+      "name": "Launch hello-world project (Buggy - use attach instead)",
```
Got it - will do... PR to follow shortly. Also, my buggy comment was totally meant to be scoped to just the hot-reloading context (I had zero issues with the standard launch approach other than not wanting to launch so often : ) )

yoDon commented:
I believe the latest commit resolves your outstanding notes with the exception of reworking the README's. It does a couple minor edit to the README's to direct the reader to use the new debug-watch commands but otherwise leaves the excess verbosity intact. I'm happy to do more work on the README's but as I mentioned in one of the other threads I'm guessing it's more efficient for you to do a first edit on one of the samples and then I can tune the rest to match your preferences (but I'm OK to just take a crack at it if that feels like I'm passing the buck to you too much).

eanders-MS commented on .vscode/launch.json:
```diff
@@ -67,6 +67,42 @@
+    },
+    {
+      "type": "node",
+      "request": "attach",
```
So, I think you only need one new configuration here, titled "Attach to running project.". I suspect the "cwd" field has no meaning when attaching. The running process already has a current working directory.

yoDon replied:
Good catch - that does mean that we're not explicitly setting the cwd for the debugger, but that's probably fine. I'll go ahead and make the change and if cwd does turn out to be needed for some workflow we can always add it back. Commit is on its way.

yoDon commented:
Launch.json updated to your suggestion of just a single attach config. Looking forward to a README commit from you. Still need to retarget onto the red branch.

yoDon commented on samples/hello-world/package.json:
```diff
@@ -23,15 +23,19 @@
     "start": "node .",
-    "debug": "node --nolazy --inspect-brk=9229 ."
+    "debug": "node --nolazy --inspect-brk=9229 .",
+    "debug-watch": "npm run build && concurrently \"npm run build-watch\" \"nodemon --nolazy --inspect=9229 .\""
```
I'm totally up for staying the course here but realistically we're into the land of small enough tweaks that are sufficiently into the personal judgement call space that it's faster and more efficient for you to just make these changes you have in mind (all seem sensible to me, none have absolute right or wrong answers so you're best arbiter of the way). 

I'm a bit concerned about the debugger message getting lost in a wall of spew (15 lines back seems enough to be at risk of getting hidden off the top of the VSCode terminal window) but as you say an appropriate bit of README guidance can go a long way. you could potentially drop the existing setTimeout snippet into an advanced readme for those who want it (since we have it) if that fits with your docs directions, or not. These are at most minor concerns and I totally get there are tradeoffs with either path.

I believe you should have edit permissions on this PR, if not let me know and I'll poke at the perms again but lets try having you finesse these fiddly bits the way you see best. And again, I'm 100% for continuing to bang on this I just don't want to be bogging down the bits that are really "add Eric's maintainer vision and hit save."

eanders-MS commented on samples/hello-world/package.json:
```diff
@@ -23,14 +23,19 @@
     "@typescript-eslint/parser": "^2.17.0",
+    "concurrently": "^5.1.0",
     "eslint": "^6.8.0",
+    "eslint-plugin-header": "^3.0.0",
```
Why is `eslint-plugin-header` needed?

yoDon replied:
Odd. No clue how that dependency ended up in package.json. Definitely not something I added intentionally. Will remove momentarily (I'll trust at the end you'll update packages-lock since you seem to default to what I'm guessing are MSFT internal mirror URLs for the npm packages).

eanders-MS replied:
We want to _avoid_ MSFT internal mirror URLs listed in package-lock.json, but sometimes they sneak in. I'll sort that out and update them.

yoDon replied:
Done, and I'm not sure where I thought I saw azure or visual studio online sources for the npm packages. Looks like everything is using the normal urls but to be safe I only updated package.json not package-lock.json.

eanders-MS replied:
OK I updated the package-lock.jsons.

eanders-MS replied:
I think "eslint-plugin-header" came in when I rebased this PR onto the red branch. Either way, I don't think we need it. Looking into removing it from `red`.

eanders-MS replied:
It's been removed from upstream already. You might want to pull from upstream for latest everything.

yoDon commented:
Just checking in - looks like you're dotting the i's and crossing the t's on the little stuff before being able to merge. Is there anything here I could/should be tackling? There's enough threading of comments that I want to make sure I'm not missing anything I could be addressing.

eanders-MS commented:
I think its done! I'm ready to merge it.
